File: submissions/0051-n-queens/solution.cpp
class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> res;
        vector<string> board(n, string(n, '.'));
        unordered_set<int> cols, diag1, diag2; // cols, row-col, row+col
        backtrack(0, n, board, res, cols, diag1, diag2);
        return res;
    }

private:
    void backtrack(int row, int n, vector<string>& board, vector<vector<string>>& res,
                   unordered_set<int>& cols, unordered_set<int>& diag1, unordered_set<int>& diag2) {
        if (row == n) {
            res.push_back(board);
            return;
        }

        for (int col = 0; col < n; ++col) {
            if (cols.count(col) || diag1.count(row - col) || diag2.count(row + col))
                continue;

            board[row][col] = 'Q';
            cols.insert(col);
            diag1.insert(row - col);
            diag2.insert(row + col);

            backtrack(row + 1, n, board, res, cols, diag1, diag2);

            // Backtrack
            board[row][col] = '.';
            cols.erase(col);
            diag1.erase(row - col);
            diag2.erase(row + col);
        }
    }
};

File: submissions/0051-n-queens/solution.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> ans = new ArrayList<>();
        char[][] board = new char[n][n];
        boolean flag = false;
        for (char[] row : board)
            Arrays.fill(row, '.');
        Queens(ans, 0, n, board);
        return ans;
    }

    public void Queens(List<List<String>> ans, int col, int n, char[][] board){
        if(col == n){
            List<String> temp = new ArrayList<>();
            for (char[] r : board)
                temp.add(new String(r));
            ans.add(temp);
            return;
        }
        for(int row = 0; row<n; row++){
            boolean flag = false;
            //upper diagonal
            if(!flag) {
                for(int i=row, j=col; i>=0 && j>=0; i--,j--){
                    if(board[i][j] == 'Q'){
                        flag = true;
                        break;
                    }
                }
            }
            //lower diagonal
            if(!flag) {
                for(int i=row, j=col; i<n && j>=0; i++,j--){
                    if(board[i][j] == 'Q'){
                        flag = true;
                        break;
                    }
                }
            }
            if(!flag){
                //horizontal
                for(int j=col; j>=0; j--){
                    if(board[row][j] == 'Q'){
                        flag = true;
                        break;
                    }
                }
            }
            

            if(flag){
                continue;
            }
            board[row][col] = 'Q';

            Queens(ans, col+1, n, board);
            board[row][col] = '.';
        }
    }
}
